src/tasks/ts/tokens.ts

```diff
@@ -0,0 +1,130 @@
+import WethNetworks from "canonical-weth/networks.json";
+import {
+  BigNumber,
+  BigNumberish,
+  Contract,
+  ContractTransaction,
+  providers,
+  Signer,
+} from "ethers";
+import { HardhatRuntimeEnvironment } from "hardhat/types";
+
+import { SupportedNetwork, isSupportedNetwork } from "./deployment";
+
+export interface Erc20Token {
+  contract: Contract;
+  symbol?: string;
+  decimals?: number;
+  address: string;
+}
+
+export interface NativeToken {
+  symbol: string;
+  decimals: number;
+  provider: providers.JsonRpcProvider;
+}
+
+export const NATIVE_TOKEN_SYMBOL: Record<
+  SupportedNetwork | "hardhat",
+  string
+> = {
+  hardhat: "ETH",
+  mainnet: "ETH",
+  rinkeby: "ETH",
+  xdai: "xDAI",
+};
+
+export const WRAPPED_NATIVE_TOKEN_ADDRESS: Record<SupportedNetwork, string> = {
+  mainnet: WethNetworks.WETH9[1].address,
+  rinkeby: WethNetworks.WETH9[4].address,
+  xdai: "0xe91D153E0b41518A2Ce8Dd3D7944Fa863463a97d",
+};
+
+export function isNativeToken(
+  token: Erc20Token | NativeToken,
+): token is NativeToken {
+  return (token as Erc20Token).contract === undefined;
+}
+
+export function displayName(token: Erc20Token | NativeToken): string {
+  if (isNativeToken(token)) {
+    return token.symbol;
+  } else {
+    return token.symbol ? `${token.symbol} (${token.address})` : token.address;
+  }
+}
+
+export async function balanceOf(
```

Comment:
I noticed that you seem to prefer free functions over instance methods. Is there any particular reason why? 

E.g. in this case I think I would have created an interface `Token` (which may have methods like transfer, balanceOf, etc) and then have both Erc20 and NativeToken implement that interface. This way you can pass around the abstract interface type in your code instead of `Erc20Token | NativeToken`.

I think at the end both is roughly equivalent but I feel free methods are more of a vintage C style (rather than more modern OOP style programming)

Reply:
You are right, a shared `Token` interface would be a good idea here.

I've had no strong opinion on which style to use when writing this. Thinking about it, I believe I used free methods to minimize the amount of changes I needed to make outside of this file to consistently integrate the new code. For the next time, I'll keep in mind to use a more aggressive style of refactoring.